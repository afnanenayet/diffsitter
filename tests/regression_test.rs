#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use libdiffsitter::{
        diff::compute_edit_script, generate_ast_vector_data, input_processing::TreeSitterProcessor,
        parse::GrammarConfig,
    };
    use std::path::PathBuf;
    use test_case::test_case;

    /// Get paths to input files for tests
    fn get_test_paths(test_type: &str, test_name: &str, ext: &str) -> (PathBuf, PathBuf) {
        let test_data_root = PathBuf::from(format!("./test_data/{test_type}/{test_name}"));
        let path_a = test_data_root.join(format!("a.{ext}"));
        let path_b = test_data_root.join(format!("b.{ext}"));
        assert!(
            path_a.exists(),
            "test data path {} does not exist",
            path_a.to_str().unwrap()
        );
        assert!(
            path_b.exists(),
            "test data path {} does not exist",
            path_b.to_str().unwrap()
        );

        (path_a, path_b)
    }

    #[test_case("short", "rust", "rs", true, true)]
    #[test_case("short", "python", "py", true, true)]
    #[test_case("short", "go", "go", true, true)]
    #[test_case("medium", "rust", "rs", true, false)]
    #[test_case("medium", "rust", "rs", false, false)]
    #[test_case("medium", "cpp", "cpp", true, true)]
    #[test_case("medium", "cpp", "cpp", false, true)]
    fn diff_hunks_snapshot(
        test_type: &str,
        name: &str,
        ext: &str,
        split_graphemes: bool,
        strip_whitespace: bool,
    ) {
        let (path_a, path_b) = get_test_paths(test_type, name, ext);
        let config = GrammarConfig::default();
        let ast_data_a = generate_ast_vector_data(path_a, None, &config).unwrap();
        let ast_data_b = generate_ast_vector_data(path_b, None, &config).unwrap();

        let processor = TreeSitterProcessor {
            split_graphemes,
            strip_whitespace,
            ..Default::default()
        };

        let diff_vec_a = processor.process(&ast_data_a.tree, &ast_data_a.text);
        let diff_vec_b = processor.process(&ast_data_b.tree, &ast_data_b.text);
        let diff_hunks = compute_edit_script(&diff_vec_a, &diff_vec_b).unwrap();

        // We have to set the snapshot name manually, otherwise there appear to be threading issues
        // and we end up with more snapshot files than there are tests, which cause
        // nondeterministic errors.
        let snapshot_name = format!("{test_type}_{name}_split_graphemes_{split_graphemes}_strip_whitespace_{strip_whitespace}");
        assert_debug_snapshot!(snapshot_name, diff_hunks);
    }
}
