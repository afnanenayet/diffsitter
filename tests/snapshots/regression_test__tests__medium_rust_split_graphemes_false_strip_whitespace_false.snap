---
source: tests/regression_test.rs
expression: snapshot_string
---
New(Line={line_index=2, entries=
[
    Entry{',', start=(2, 26), end=(2, 26)}
]}
)
New(Line={line_index=20, entries=
[
    Entry{'is_naked_fn', start=(20, 4), end=(20, 4)}
]}
)
Old(Line={line_index=17, entries=
[
    Entry{'is_naked', start=(17, 7), end=(17, 7)}
]}
)
New(Line={line_index=42, entries=
[
    Entry{',', start=(42, 19), end=(42, 19)}
]}
)
New(Line={line_index=59, entries=
[
    Entry{'bleat', start=(59, 4), end=(59, 4)}
]}
)
Old(Line={line_index=53, entries=
[
    Entry{'talk', start=(53, 7), end=(53, 7)}
]}
)
New(Line={line_index=61, entries=
[
    Entry{'{} beats briefly... {}', start=(61, 12), end=(61, 12)}
]}
)
Old(Line={line_index=55, entries=
[
    Entry{'{} pauses briefly... {}', start=(55, 18), end=(55, 18)}
]}
)
New(Line={line_index=67, entries=
[
    Entry{'ed', start=(67, 9), end=(67, 9)}
    Entry{'Logan', start=(67, 34), end=(67, 34)}
]}
)
Old(Line={line_index=61, entries=
[
    Entry{'dolly', start=(61, 12), end=(61, 12)}
    Entry{'Dolly', start=(61, 40), end=(61, 40)}
]}
)
New(Line={line_index=70, entries=
[
    Entry{'ed', start=(70, 1), end=(70, 1)}
    Entry{'bleat', start=(70, 4), end=(70, 4)}
]}

Line={line_index=71, entries=
[
    Entry{'ed', start=(71, 1), end=(71, 1)}
]}

Line={line_index=72, entries=
[
    Entry{'ed', start=(72, 1), end=(72, 1)}
    Entry{'bleat', start=(72, 4), end=(72, 4)}
]}
)
Old(Line={line_index=64, entries=
[
    Entry{'dolly', start=(64, 4), end=(64, 4)}
    Entry{'talk', start=(64, 10), end=(64, 10)}
]}

Line={line_index=65, entries=
[
    Entry{'dolly', start=(65, 4), end=(65, 4)}
]}

Line={line_index=66, entries=
[
    Entry{'dolly', start=(66, 4), end=(66, 4)}
    Entry{'talk', start=(66, 10), end=(66, 10)}
]}
)
